File: sdks/sdk-typescript/src/webhook.ts
import { createVerify } from 'node:crypto'
import { OpenPayError } from './errors.js'
import type { WebhookEvent } from './types.js'

/**
 * Verify an incoming webhook signature using ED25519.
 * The Open Pay webhook service signs payloads with ED25519.
 *
 * @param payload - The raw request body string
 * @param headers - The request headers containing signature, timestamp, event, and ID
 * @param publicKey - The ED25519 public key (base64-encoded, from GET /v1/webhooks/public-key)
 * @returns The parsed webhook event
 */
export function verifyWebhookSignature(
  payload: string,
  headers: {
    'x-webhook-signature'?: string
    'x-webhook-timestamp'?: string
    'x-webhook-event'?: string
    'x-webhook-id'?: string
  },
  publicKey: string,
): WebhookEvent {
  const signature = headers['x-webhook-signature']
  const timestamp = headers['x-webhook-timestamp']
  const event = headers['x-webhook-event']
  const id = headers['x-webhook-id']

  if (!signature || !timestamp || !event || !id) {
    throw new OpenPayError('Missing webhook signature headers')
  }

  // Verify ED25519 signature: message = timestamp_bytes + payload_bytes
  const message = Buffer.concat([
    Buffer.from(timestamp, 'utf8'),
    Buffer.from(payload, 'utf8'),
  ])

  const sigBytes = Buffer.from(signature, 'base64')
  const pubKeyBytes = Buffer.from(publicKey, 'base64')

  const verify = createVerify('ed25519')
  verify.update(message)

  // ed25519 keys need to be in DER format for Node.js crypto
  const derPrefix = Buffer.from('302a300506032b6570032100', 'hex')
  const derKey = Buffer.concat([derPrefix, pubKeyBytes])

  const isValid = verify.verify(
    { key: derKey, format: 'der', type: 'spki' },
    sigBytes,
  )

  if (!isValid) {
    throw new OpenPayError('Invalid webhook signature')
  }

  // Check timestamp freshness (5 minutes)
  const ts = parseInt(timestamp, 10)
  const now = Date.now()
  if (Math.abs(now - ts) > 5 * 60 * 1000) {
    throw new OpenPayError('Webhook timestamp too old')
  }

  return {
    id,
    event,
    timestamp,
    data: JSON.parse(payload),
  }
}

File: sdks/sdk-python/openpay/webhook.py
"""Webhook signature verification using ED25519."""

import json
import time
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
from cryptography.exceptions import InvalidSignature
import base64

from openpay.errors import OpenPayError
from openpay.types import WebhookEvent


def verify_webhook_signature(
    payload: str,
    signature_b64: str,
    timestamp: str,
    event: str,
    webhook_id: str,
    public_key_b64: str,
    max_age_seconds: int = 300,
) -> WebhookEvent:
    """Verify an incoming webhook signature using ED25519.

    Args:
        payload: The raw request body string
        signature_b64: The X-Webhook-Signature header (base64)
        timestamp: The X-Webhook-Timestamp header
        event: The X-Webhook-Event header
        webhook_id: The X-Webhook-ID header
        public_key_b64: The ED25519 public key (base64)
        max_age_seconds: Maximum age of the webhook in seconds (default 300)

    Returns:
        Parsed WebhookEvent

    Raises:
        OpenPayError: If verification fails
    """
    if not all([payload, signature_b64, timestamp, event, webhook_id, public_key_b64]):
        raise OpenPayError("Missing webhook signature parameters")

    # Check timestamp freshness
    try:
        ts = int(timestamp)
        if abs(time.time() * 1000 - ts) > max_age_seconds * 1000:
            raise OpenPayError("Webhook timestamp too old")
    except ValueError:
        raise OpenPayError("Invalid webhook timestamp")

    # Verify signature
    try:
        pub_bytes = base64.b64decode(public_key_b64)
        sig_bytes = base64.b64decode(signature_b64)
        message = timestamp.encode() + payload.encode()

        public_key = Ed25519PublicKey.from_public_bytes(pub_bytes)
        public_key.verify(sig_bytes, message)
    except InvalidSignature:
        raise OpenPayError("Invalid webhook signature")
    except Exception as e:
        raise OpenPayError(f"Webhook verification failed: {e}")

    return WebhookEvent(
        id=webhook_id,
        event=event,
        timestamp=timestamp,
        data=json.loads(payload),
    )
